### PR TITLE
Adds a new traitor theft objective, stealing the Blackbox.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
No, not the TF2 one.

## About The Pull Request

You ever been fixing the telecomms array, and noticed  the blackbox recorder? It's that nice looking machine that just sits there, spinning, probably doing something with telecoms like recording messages or something or another.

Sorry to burst your bubble, but it does literally nothing. It just sits there, spinning. Forever.

Not anymore!

Now, the blackbox recorder contains the one and only, blackbox. Like a real blackbox, this thing is indestructible as hell. It's nature, unknown, it's defense mechanism, unseen, but one thing is for sure, it holds data.

A lot of data.
image

A lot of data that the syndicate **WANTS**.
...Assuming that they can figure out how to crack this thing wide open, but first they need their agents to steal it.

## Why It's Good For The Game

This is an idea I've been sitting on since the antag freeze started, so that the blackbox could serve as a real and true high fidelity item on the station. If nothing else, it's difficult to get access to (Engineers, atmos techs, and the CE is blacklisted from this objective), it's a bulky item so it's difficult to conceal and hide, and the challenge is finding a discreet way to get it off the station without it looking too suspicious. Plus, it helps a piece of flavor on the station have a slightly more concrete purpose.

## Changelog
:cl:
add: The Syndicate has recently begun targeting a recently declassified piece of Nanotrasen hardware, the blackbox, located within the blackbox recorder on each station's telecommunication's array.
/:cl:

### CODE REVIEW

**Comment by cacogen** [comment]
I think it used to prevent stats being recorded for that round if it was destroyed, back in the days when errorage was running the stats page? I'm assuming it doesn't do that anymore because it seems like a dumb idea to let game stats be skewed by an IC item?

**Comment by TheVekter** [comment]
It's bulky, so does it fit in a Bag of Holding?

**Comment by ATH1909** [comment]
So many people are gonna redtext with this objective because they weren't buckled into a seat when the shuttle arrived at Centcom.

**Comment by Buggy123** [comment]
>Engineers, atmos techs, and the CE is blacklisted from this objective

While this isn't unreasonable, I don't know if you should do this considering all of these people can get the steal supermatter sliver objective.

**Comment by necromanceranne** [comment]
I desire the cube

**Comment by TheVekter** [comment]
> 
> 
> > Engineers, atmos techs, and the CE is blacklisted from this objective
> 
> While this isn't unreasonable, I don't know if you should do this considering all of these people can get the steal supermatter sliver objective.

Supermatter sliver still has a significant amount of danger to it. You get dosed hard with rads and if you don't have grav boots or you're a dummy you get dusted.

**Comment by PKPenguin321** [comment]
Should probably give it some semblance of a purpose so the crew can't just hide it roundstart

**Comment by Buggy123** [comment]
Big brain: Make it actually store the comm logs for the entire round.

**Comment by HomocidlePancakes** [comment]
>Should probably give it some semblance of a purpose so the crew can't just hide it roundstart

I doubt anyone will be rushing in to comms to hide the the blackbox, and even if somebody does, they are the kind that space all high-risk items roundstart.

**Comment by necromanceranne** [comment]
Which we have rules against.

**Comment by PKPenguin321** [comment]
There's no rule against not giving a shit if an antag takes an item with no functionality either
This is arguably less interesting than the slime core steal objective

**Comment by necromanceranne** [comment]
Well the rule specifically is more 'you are not obligated to intervene against an antagonist' though with some exceptions given to security with regards to people being attacked. So the issue is more that people are not obligated to stop you or validhunt or scrutinize, which I don't see changing any time soon nor would we want to see that changed.

**Comment by Dorsisdwarf** [comment]
Blackbox being destroyed should wipe the radio records from comms maybe?

**Comment by HomocidlePancakes** [comment]
Wait, wasn’t there a PR with a t-comms high-risk item that had actual functionality?

**Comment by Dennok** [comment]
> Blackbox being destroyed should wipe the radio records from comms maybe?

Radio records keeped in servers and you can see it in console. But it store a maximum of 100 logs.

**Comment by fludd12** [comment]
Everyone stops earning money if the black box is removed, so people have a reason to make it not stolen but also not space it

**Comment by Anonmare** [comment]
The cube will save us

File: code/game/machinery/telecomms/machines/message_server.dm [hunk 1]
```diff
@@ -15,7 +15,60 @@
...
 	armor = list("melee" = 25, "bullet" = 10, "laser" = 10, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 50, "acid" = 70)
+	var/obj/item/stored
 
+/obj/machinery/blackbox_recorder/Initialize()
+	. = ..()
+	stored = new /obj/item/blackbox(src)
+
+/obj/machinery/blackbox_recorder/attack_hand(mob/living/user)
+	. = ..()
+	if(stored)
```
Comment (Fikou) [inline code/game/machinery/telecomms/machines/message_server.dm]: should probably be a do_after?

**Comment by itseasytosee** [comment]
I don't like that the object serves no purpose other than being an objective. there are very few items like this being the captains badge, which has nice flavor. and the plutonium core which has  fun mechanic and disables the nuke. 
Also having a bulky item objective is unreasonable.

**Comment by Cyberboss** [comment]
I think this could be good, but only if the crew was required to recover the blackbox during evacuation as well.

**Comment by nfreader** [comment]
I’ve got a design document for a blackbox item like this:

# The Datacore
A new machine with a removable *Datacore*, which stores all the station’s security, medical, and financial records. The Datacore is a high-profile traitor theft objective.

The datacore will handle:
• Crew roster
• Medical records
• Security records
• Financial transactions (ledger)
• Research 

## Required code changes
- + Data core machine
	- + Data core item
- Migrating/copying all data to the datacore, handling loss events
	- Or just using the datacore as a proxy/boolean check (does this item exist in this machine?)
- Update consoles and systems to require a working datacore in order to function 
- Alert functions for missing or disabled datacore
- Minor crew success for leaving with the datacore?

**Comment by itseasytosee** [comment]
Having a crew objective every round will lead to people just not caring like evacing the AI

**Comment by AWalton** [comment]
Making it Bulky quite possibly makes this the most difficult objective you could get.

**Comment by fludd12** [comment]
> Making it Bulky quite possibly makes this the most difficult objective you could get.

I agree, but I don't think that's a bad thing. Difficult objectives can be really fun, too. I think coupled with @nfreader's "Datacore" idea, it'd make for a very interesting objective.

I think, rather than making it a crew victory to have the datacore intact, instead make having certain amounts of research done or records available add to the end of round score - and without the datacore, they aren't added, and the loss of the datacore is marked as a somewhat negative mod to the score.

Also, future plans, now that I think about it; what if the previous round's score somehow adjusted budgets / the station in some way? It'd give a reason for the crew to try to maximize their score.

**Comment by trollbreeder** [comment]
> Also, future plans, now that I think about it; what if the previous round's score somehow adjusted budgets / the station in some way? It'd give a reason for the crew to try to maximize their score.

Rounds should be completely separate from eachother in the code, except in non-altering ways like how poly remembers things said to him.

File: code/game/machinery/telecomms/machines/message_server.dm [hunk 2]
```diff
@@ -15,7 +15,60 @@
...
+	if(istype(I, /obj/item/blackbox))
+		if(HAS_TRAIT(I, TRAIT_NODROP) || !user.transferItemToLoc(I, src))
+			to_chat(user, "<span class='warning'>[I] is stuck to your hand!</span>")
+			return
+		user.visible_message("<span class='notice'>[user] clicks [I] into [src]!</span>", \
+		"<span class='notice'>You press the device into [src], and it clicks into place. The tapes begin spinning again.</span>")
+		playsound(src, 'sound/machines/click.ogg', 50, TRUE)
+		stored = I
+		update_icon()
+		return ..()
```
Comment (ShizCalev) [inline code/game/machinery/telecomms/machines/message_server.dm]: if both these return ..(), might as well just remove this

**Comment by itseasytosee** [comment]
Not a single change or comment between massive amounts of feedback and eventual merge. Great

**Comment by ArcaneMusic** [comment]
Look, I'd be very glad to see @nfreader 's changes come forward, but It would requires much larger scope than I had originally (Large, difficult to steal, mysterious box). That being said I had to deal with several exams since opening this PR and this next week, so I was time limited since then. Still, there's a few things I'd be fine with expanding on further (As well as see some feedback on how difficult it is to steal the cube).

**Pull request merged**